tools/tsflower

```diff
@@ -147,6 +147,24 @@ run_only()
     # TODO get list of packages from data... better yet, make it
     #   one tsflower command, reading a TsFlower config file
 
+    package=expo-notifications
+    run_on_package "${package}"
+    rm -rf "${rootdir}"/types/"${package}"/node_modules/
+    files_to_delete=$(
+        find "${rootdir}"/types/"${package}" -type f \
+            -not -name 'NotificationPermissions*' \
+            -not -path "${rootdir}"/types/"${package}"/index.js.flow \
+            -not -path "${rootdir}"/types/"${package}"/build/index.js.flow \
+    )
+    # We only use one tiny piece of this library: querying whether the user
+    # has granted notification permissions.
+    rm $files_to_delete
```

Comment:
This will break if `${rootdir}` has a space in it:
* The `files_to_delete` variable gets set to just a string with the output of `find`, where `find` has used newlines to separate the filenames.
* Then in the command `rm $files_to_delete`, the `files_to_delete` variable gets substituted in… and *split* on space, tab, and newline to produce the list of arguments that get passed to `rm`.
* So if the user/developer's worktree is at a path like `/home/doe/My Documents/zulip-mobile/`, then the arguments to `rm` will include `/home/doe/My` and `Documents/zulip-mobile/types/expo-notifications/build/BadgeModule.js.flow` etc.
* Usually this only means the `rm` will print a bunch of errors and fail to remove what it was intended to remove.

  Unless there actually is a file named `/home/doe/My`, that is. Hopefully that wasn't something the user wanted to keep. :'-/

This sort of bug is classic in shell scripting. It's the biggest reason why you see double-quotes `"…"` all over shell scripts that I or Anders write — because that's what instructs the shell to take the strings as plain data and not re-parse them like this.

Here, though, just `rm "${files_to_delete}"` won't work either — the splitting is serving a purpose of separating the filenames that `find` gave us.

There are several strategies to fix a script like this:
* Use nulls as delimiters instead of newlines. The null byte `\x00` is the only byte that can't appear in a path, which makes it a legitimate delimiter.

  Many tools like `find` will take an option to do this. In the case of `find`, add `-print0` at the end (there's an implicit `-print`, so this replaces that), and then pipe to `xargs -0`. Here's a pair of examples from `man find`:

       find /tmp -name core -type f -print | xargs /bin/rm -f

       Find files named core in or below the directory /tmp and  delete  them.
       Note  that  this  will work incorrectly if there are any filenames con‐
       taining newlines, single or double quotes, or spaces.

       find /tmp -name core -type f -print0 | xargs -0 /bin/rm -f

       Find files named core in or below the directory /tmp and  delete  them,
       processing  filenames  in  such a way that file or directory names con‐
       taining single or double quotes, spaces or newlines are correctly  han‐
       dled.

* Have a single tool do the job, without any shell-style communication between tools. `find` can do this with `-delete`. For this example, one might also try writing a `git clean` command line.

* Arrange to deal only with filenames that come from a more controlled source, and then hope that source doesn't throw us a surprising filename after all.

  For example, here we could rely on the `cd "${rootdir}"` at top (or write `(cd "${rootdir}" && …)` locally, and do `find ./types/"${package}" …`. We'd be counting on the package not turning out to contain a filename with whitespace in it. This is nearly always a successful assumption for software trees (at least for the kinds of software I'm accustomed to, including NPM packages — might not hold for the macOS- or Windows-native ecosystems). Still not great to rely on, though.

  This strategy is an improvement over relying also on assumptions about the *user's* choice of paths, but it's hacky at best; the others are much preferred.

Reply:
> * Unless there actually is a file named `/home/doe/My`, that is. Hopefully that wasn't something the user wanted to keep. :'-/

!! Indeed!

Reply:
Thanks for all of this explanation!

Reply:
Here's something I came up with that seems to work:

```bash
    # We only use one tiny piece of this library: querying whether the user
    # has granted notification permissions. Remove everything else.
    git clean --force -d --quiet \
        --exclude "${rootdir}"/types/"${package}"/NotificationPermissions* \
        --exclude "${rootdir}"/types/"${package}"/index.js.flow \
        --exclude "${rootdir}"/types/"${package}"/build/index.js.flow \
        -- "${rootdir}"/types/"${package}"
```

Although this won't be necessary in a revision that has us use our own new native modules instead of `expo-notifications`.

Reply:
Cool. Yeah, that looks like a good solution.

I believe your original `find` command, with `-delete` added at the end, and without the `files_to_delete` variable or the subsequent `rm` command, would also do it.

Reply:
I remember trying a few different things yesterday with `find -delete` that didn't end up working, but I'm not sure why.